entry/src/block_component.rs

```diff
@@ -0,0 +1,2149 @@
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Parent Slot                  (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Parent Block ID             (32 bytes)  │
+/// └─────────────────────────────────────────┘
+/// ```
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct ParentReadyUpdateV1 {
+    pub new_parent_slot: Slot,
+    pub new_parent_block_id: Hash,
+}
+
+// ============================================================================
+// BlockComponent Implementation
+// ============================================================================
+
+impl Default for BlockComponent {
+    fn default() -> Self {
+        Self::Entries(Vec::new())
+    }
+}
+
+impl BlockComponent {
+    /// Maximum number of entries allowed in a block component.
+    ///
+    /// TODO(karthik): lower this to a reasonable value.
+    const MAX_ENTRIES: usize = u32::MAX as usize;
+
+    /// Creates a new block component with entries.
+    ///
+    /// # Errors
+    /// Returns an error if the entries vector is empty or exceeds the maximum allowed size.
+    pub fn new_entries(entries: Vec<Entry>) -> Result<Self, String> {
+        if entries.is_empty() {
+            return Err("BlockComponent with entries cannot be empty".to_string());
+        }
+        Self::validate_entries_length(entries.len())?;
+        Ok(Self::Entries(entries))
+    }
+
+    /// Creates a new block component with a special marker.
+    pub const fn new_block_marker(marker: VersionedBlockMarker) -> Self {
+        Self::BlockMarker(marker)
+    }
+
+    /// Returns a slice of entries in this component.
+    pub fn entries(&self) -> &[Entry] {
+        match self {
+            Self::Entries(entries) => entries,
+            Self::BlockMarker(_) => &[],
+        }
+    }
+
+    /// Returns the special marker if present.
+    pub const fn marker(&self) -> Option<&VersionedBlockMarker> {
+        match self {
+            Self::Entries(_) => None,
+            Self::BlockMarker(marker) => Some(marker),
+        }
+    }
+
+    /// Returns true if this component contains entries.
+    pub const fn is_entries(&self) -> bool {
+        matches!(self, Self::Entries(_))
+    }
+
+    /// Returns true if this component contains a special marker.
+    pub const fn is_marker(&self) -> bool {
+        matches!(self, Self::BlockMarker(_))
+    }
+
+    /// Validates that the entries length is within bounds.
+    fn validate_entries_length(len: usize) -> Result<(), String> {
+        if len >= Self::MAX_ENTRIES {
+            Err(format!(
+                "BlockComponent entries length {len} exceeds maximum {}",
+                Self::MAX_ENTRIES
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Serializes to bytes.
+    ///
+    /// # Errors
+    /// Returns a bincode error if serialization fails or if validation fails.
+    pub fn to_bytes(&self) -> Result<Vec<u8>, bincode::Error> {
```

Comment:
nit: instead of every `pub fn` returning a different error, do you think we should define `BlockComponentError`?